Fix delegate cgroup logs filtering

qdeslandes opened the pull request:
For services using delegation, `systemd-journald` wouldn't be able to filter the subcgroups logs as it would read the extended attributes from the subcgroup itself, instead of the unit's cgroup (see #26024).

Fixes it by introducing a function to get the service's cgroup path from a given cgroup path. 

This change has been tested with `Delegate=yes` and the unit's attributes are properly used by the subcgroup, also I've used the same test cases are `cg_path_get_unit` to ensure the behaviour is similar.

Fixes #26024 

<!-- devel-freezer = {"comment-id":"1418939559","freezing-tag":"v253-rc2"} -->

bluca commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,35 @@ int cg_path_get_unit(const char *path, char **ret) {
+        if (!unit_name_is_valid(unit_name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
+                return -ENXIO;
+
+        *ret = strdup(path_copy);
```
Oom check

bluca commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,37 @@ int cg_path_get_unit(const char *path, char **ret) {
+        if (!unit_name_is_valid(unit_name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
+                return -ENXIO;
+
+        *ret = strdup(path_copy);
```
This will overwrite ret when it fails, which we avoid as per coding style, use an auxiliary local variable until after the last error check

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,37 @@ int cg_path_get_unit(const char *path, char **ret) {
+        if (unit_name[0] == '_')
+                /* Move unit_name one byte left to override '_'. This will
+                 * also move the NUL character. */
+                memmove(unit_name, &unit_name[1], unit_name_len);
```
please don#t write this manually, just use cg_unescape()

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,37 @@ int cg_path_get_unit(const char *path, char **ret) {
+
+        *ret = strdup(path_copy);
+        if (!*ret)
+                return log_oom();
```
use just `return -ENOMEM` here.

(we want that functions either log about all errors or about none [in which case the error should be propagated so that caller may log], and library functions are typically the ones where logging should be left to the caller. hence use ENOMEM here, without logging)

bluca commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,38 @@ int cg_path_get_unit(const char *path, char **ret) {
+
+        /* If unit_name is escaped, overwrite it in place with the unespaced
+         * version. memmove() doesn't mind overlapping memory areas. */
+        memmove(unit_name, cg_unescape(unit_name), unit_name_len);
```
why is the memmove needed here? just pass cg_unescape to unit_name_is_valid?

qdeslandes replied:
We store a copy of `path` into `path_copy` so we can modify it: 
- Add `\0` after the unit's name.
- Remove escape character, if any.

Then, the modified `path_copy` is strdup()'ed into the return value. `memmove()` is used to ensure the escaping character is removed from `path_copy`.

qdeslandes commented:
Unit test failure affects `systemd:test / test-time-util`, by quickly looking at the code, I'm keen to think it's unrelated.

bluca commented:
> Unit test failure affects `systemd:test / test-time-util`, by quickly looking at the code, I'm keen to think it's unrelated.

https://github.com/systemd/systemd/issues/26172

keszybz commented:
LGTM.

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,38 @@ int cg_path_get_unit(const char *path, char **ret) {
+        *ret = unit_path;
+
+        return 0;
+}
```
this looks broken.

You unescape the name, and leave that in the path, so you'll return a path that doesn't actually exist if it was escaped? that's wrong.

poettering commented on src/test/test-cgroup-util.c:
```diff
@@ -63,6 +63,30 @@ TEST(path_get_unit) {
+        check_p_g_u_p("/system.slice/system-waldo.slice/foobar.service/sdfdsaf", 0, "/system.slice/system-waldo.slice/foobar.service");
+        check_p_g_u_p("/system.slice/system-waldo.slice/_cpu.service/sdfdsaf", 0, "/system.slice/system-waldo.slice/cpu.service");
+        check_p_g_u_p("/user.slice/user-1000.slice/user@1000.service/server.service", 0, "/user.slice/user-1000.slice/user@1000.service");
+        check_p_g_u_p("/user.slice/user-1000.slice/user@.service/server.service", -ENXIO, NULL);
```
this really should have escaped examples

yuwata commented:
@bluca I think the filtering feature is merged after v252. Is this really necessary to be backported??

bluca commented:
> @bluca I think the filtering feature is merged after v252. Is this really necessary to be backported??

ah if it's a new thing then no need

qdeslandes commented:
I misunderstood escaping, I think it's fixed now. To summarise changes:
- `cg_path_get_unit_path()` will get the unit's path from a valid cgroup path.
- From a given cgroup path, once we have skipped the slices, ensure with have a valid unit name before the next `/` (if any). If so, return the path until the next `/`, otherwise return an error.
- Added more unit tests, especially for escaped paths. 

Once we skip slices in a path, should we expect nested `systemd` constructs? For example: `/system.slice/user.scope/test.service` or something similar? I assumed we should not, but I'm not 100%. I added tests with nested constructs, even if it's not supposed to happen, we know at least how `cg_path_get_unit_path()` would behave.

yuwata commented on src/journal/journald-client.c:
```diff
@@ -46,16 +46,20 @@ static int client_parse_log_filter_nulstr(const char *nulstr, size_t len, Set **
 
```
commit message: "Fixed by unit" ??? Maybe, "Fixed by using"?

yuwata reviewed (approved):
LGTM. Though, I am not familiar with cgroup delegation.

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,34 @@ int cg_path_get_unit(const char *path, char **ret) {
 }
 
+int cg_path_get_unit_path(const char *path, char **ret) {
+        char *path_copy, *unit_name, *unit_path;
```
please make unit_name const

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,34 @@ int cg_path_get_unit(const char *path, char **ret) {
+
+        path_copy = strdupa_safe(path);
+
+        unit_name = (char *)skip_slices(path_copy);
```
then you don't have to cast the "const" away here

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,34 @@ int cg_path_get_unit(const char *path, char **ret) {
+
+        unit_name[unit_name_len] = '\0';
+
+        if (!unit_name_is_valid(unit_name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
```
now you dropped the unescaping entirely. 

you should do the unescaping here, but only for the validation step, and not modify the path string.

i.e. it's sufficient to change this line here:

```c
if (!unit_name_is_valid(cg_unescape(unit_name), UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
```

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1198,6 +1198,34 @@ int cg_path_get_unit(const char *path, char **ret) {
+        assert(path);
+        assert(ret);
+
+        path_copy = strdupa_safe(path);
```
instead of copying the string twice (once here via strdupa_alloc() on the stack, and once below via strdup() on the heap). Just allocate it here onto the heap right away.

poettering replied:
i.e.

```c
_cleanup_free_ char *path_copy = strdup(path);
if (!path_copy)
        return -ENOMEM;

unit_name = (char*) skip_slices(path_copy);
unit_name[strcspn(unit_name, "/")] = 0;

if (!unit_name_is_valid(cg_unescape(unit_name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
        return -ENXIO;

*ret = TAKE_PTR(path_copy);
return 0;
``

poettering commented:
lgtm, but ci fell over its own feet. Any chance you can rebase/repush to trigger the Ci again?

bluca commented:
no need, I'll just retrigger the failed ones

keszybz reviewed (commented):
LGTM.

yuwata commented:
May not need to address in this PR, but please also update TEST-02-JOURNAL to check the issue is actually fixed.

bluca commented on test/units/delegated_cgroup_filtering_payload.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
```
`# SPDX-License-Identifier: LGPL-2.1-or-later` in both scripts please
